src/net_processing.cpp

```diff
@@ -468,6 +468,44 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     return &it->second;
 }
 
+/**
+ * Data stucture for an individual peer. This struct is not protected by
+ * cs_main since it does not contain validation-critical data.
+ *
+ * Memory is owned by shared pointers and this object is destructed when
+ * the refcount drops to zero (in FinalizeNode()).
+ *
+ * TODO: move most members from CNodeState to this structure.
+ * TODO: move remaining application-layer data members from CNode to this structure.
+ */
+struct Peer {
+    /** Same id as the CNode object for this peer */
+    NodeId m_id;
+
+    Peer(NodeId id) : m_id(id) {}
+};
+
+using PeerRef = std::shared_ptr<Peer>;
+
+/**
+ * Map of all Peer objects, keyed by peer id. This map is protected
+ * by the global g_peer_mutex. Once a shared pointer reference is
+ * taken, the lock may be released. Individual fields are protected by
+ * their own locks.
+ */
+Mutex g_peer_mutex;
+static std::map<NodeId, PeerRef> g_peer_map GUARDED_BY(g_peer_mutex);
+
+/** Get a shared pointer to the Peer object.
+ *  May return nullptr if the Peer object can't be found. */
+static PeerRef GetPeer(NodeId id) {
+    PeerRef ret;
+    LOCK(g_peer_mutex);
+    auto it = g_peer_map.find(id);
+    if (it != g_peer_map.end()) ret = it->second;
+    return ret;
```

Comment:
nit in same commit:

Can be written with one less line ;)

```suggestion
    if (it != g_peer_map.end()) return it->second;
    return {}; // ( or return nullptr; )
```

Reply:
This may be a rare case where a ternary makes things clearer.
```suggestion
    return (it != g_peer_map.end()) ? it->second : nullptr;
```

Reply:
Done with Troy's ternary. Thanks!